src/env-inl.h

```diff
@@ -217,6 +231,41 @@ void AsyncHooks::clear_async_id_stack() {
   fields_[kStackLength] = 0;
 }
 
+inline void AsyncHooks::AddContext(v8::Local<v8::Context> ctx) {
+  ctx->SetPromiseHooks(
+    js_promise_hooks_[0].IsEmpty() ?
+      v8::Local<v8::Function>() :
+      PersistentToLocal::Strong(js_promise_hooks_[0]),
+    js_promise_hooks_[1].IsEmpty() ?
+      v8::Local<v8::Function>() :
+      PersistentToLocal::Strong(js_promise_hooks_[1]),
+    js_promise_hooks_[2].IsEmpty() ?
+      v8::Local<v8::Function>() :
+      PersistentToLocal::Strong(js_promise_hooks_[2]),
+    js_promise_hooks_[3].IsEmpty() ?
+      v8::Local<v8::Function>() :
+      PersistentToLocal::Strong(js_promise_hooks_[3]));
```

Comment:
Can you add these in `AsyncHooks::MemoryInfo` (in env.cc) as well?

Reply:
Did this (though I'm not sure I did it correctly) in latest fixup commit.

Reply:
Sorry, yeah, I should probably have provided more context here. `MemoryInfo` is for tracking strong keep-alive relations between heap objects – so we’re not interested in the contexts, because those are weak handles, and it would be misleading to act like our async_hooks tracking is keeping them in memory. However, it is keeping the hook functions alive, and that’s actually useful information in a heap dump because it makes it much easier to identify them.

So what we want is `tracker->TrackField("js_promise_hooks", js_promise_hooks_);`, which doesn’t work out of the box because `js_promise_hooks_` is a C array – we could add support for C arrays, but it’s fairly unusual to work with them anyway, and `std::array` does work because it follows the format of a standard C++ container, so the following would be what we want here:

```diff
diff --git a/src/env.cc b/src/env.cc
index 5ee489fafdf4..042b7243ac10 100644
--- a/src/env.cc
+++ b/src/env.cc
@@ -1164,10 +1164,7 @@ void AsyncHooks::MemoryInfo(MemoryTracker* tracker) const {
   tracker->TrackField("async_ids_stack", async_ids_stack_);
   tracker->TrackField("fields", fields_);
   tracker->TrackField("async_id_fields", async_id_fields_);
-  tracker->TrackFieldWithSize(
-      "contexts", contexts_.size() * sizeof(v8::Global<Context>));
-  tracker->TrackFieldWithSize(
-      "js_promise_hooks", 4 * sizeof(v8::Global<Function>));
+  tracker->TrackField("js_promise_hooks", js_promise_hooks_);
 }
 
 void AsyncHooks::grow_async_ids_stack() {
diff --git a/src/env.h b/src/env.h
index 08d7b4d59ef6..ba7cde0d447f 100644
--- a/src/env.h
+++ b/src/env.h
@@ -781,7 +781,7 @@ class AsyncHooks : public MemoryRetainer {
 
   std::vector<v8::Global<v8::Context>> contexts_;
 
-  v8::Global<v8::Function> js_promise_hooks_[4];
+  std::array<v8::Global<v8::Function>, 4> js_promise_hooks_;
 };
 
 class ImmediateInfo : public MemoryRetainer {
```

Reply:
Ah, ok, gotcha. Thanks! I've changed that now.